### PR TITLE
[consensus] Remove `Coordinator`

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Removes the `Coordinator` trait from `commonware-resolver` in favor of a new `PeerSetProvider` trait implemented by the oracles in `commonware-p2p`.

Also updates the `commonware-reshare` example to dynamically change the peer set as epochs transition. At any given time, the peer set includes:
1. The players in `current_epoch + 1`, to allow for them to have 2 epochs to resync before their participation in the committee.
2. The players/dealers in `current_epoch`.

### CODE REVIEW

File: p2p/src/authenticated/discovery/actors/tracker/ingress.rs [hunk 1]
```diff
@@ -12,11 +12,26 @@ use commonware_utils::set::Ordered;
...
+    // ---------- Used by peer set provider ----------
+    /// Fetch the peer set at a given index.
+    PeerSet {
+        /// The index of the peer set to fetch.
+        index: u64,
+        /// One-shot channel to send the peer set.
+        responder: oneshot::Sender<Option<Ordered<C>>>,
+    },
+    /// Fetch the latest peer set index.
+    PeerSetId {
```
Comment (patrick-ogrady) [inline p2p/src/authenticated/discovery/actors/tracker/ingress.rs]: nit: `LatestPeerSet`

File: p2p/src/authenticated/discovery/actors/tracker/ingress.rs [hunk 2]
```diff
@@ -232,6 +247,29 @@ impl<C: PublicKey> Oracle<C> {
     }
 }
 
+impl<C: PublicKey> crate::PeerSetProvider for Oracle<C> {
+    type PublicKey = C;
+
+    async fn peer_set_id(&mut self) -> Option<u64> {
```
Comment (patrick-ogrady) [inline p2p/src/authenticated/discovery/actors/tracker/ingress.rs]: nit: `latest_peer_set`

Comment (patrick-ogrady) [reply p2p/src/authenticated/discovery/actors/tracker/ingress.rs]: (I think `id` is implied by return type)

File: p2p/src/lib.rs [hunk 3]
```diff
@@ -66,6 +67,27 @@ pub trait Receiver: Debug + Send + 'static {
...
+/// Interface for registering new peer sets as well as fetching an ordered list of connected peers, given a set id.
+pub trait PeerSetManager: Debug + Clone + Send + 'static {
+    /// Public key type used to identify peers.
+    type PublicKey: PublicKey;
+
+    /// The type for the peer set in registration.
+    type Peers;
+
+    /// Register a new ordered set of peers for a given ID.
+    fn register(&mut self, id: u64, peers: Self::Peers) -> impl Future<Output = ()> + Send;
```
Comment (patrick-ogrady) [inline p2p/src/lib.rs]: Only now realizing we use `register` for peer sets and for channels everywhere. I think we should delineate between the two.

Let's call this register `update` and keep the channel `register` as is.

File: p2p/src/simulated/ingress.rs [hunk 4]
```diff
@@ -205,6 +225,25 @@ pub struct Control<P: PublicKey> {
     sender: mpsc::UnboundedSender<Message<P>>,
 }
 
+impl<P: PublicKey> Control<P> {
+    /// Register the communication interfaces for the peer over a given [Channel].
+    pub async fn register_comms(
```
Comment (patrick-ogrady) [inline p2p/src/simulated/ingress.rs]: See comment about keeping this `register` and changing peer set changes to `update`

File: p2p/src/simulated/network.rs [hunk 5]
```diff
@@ -169,9 +173,24 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
         }
 
         match message {
-            ingress::Message::Register {
-                public_key,
+            ingress::Message::Register { peer_set, peers } => {
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs]: nit: require monotonically increasing (and log if not) like other p2p

File: p2p/src/simulated/network.rs [hunk 6]
```diff
@@ -169,9 +173,24 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
...
+                for public_key in peers {
+                    if !self.peers.contains_key(&public_key) {
+                        let peer = Peer::new(
+                            self.context.with_label("peer"),
+                            public_key.clone(),
+                            self.get_next_socket(),
+                            self.max_size,
+                        );
+                        self.peers.insert(public_key.clone(), peer);
+                    }
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs:232]: We need to drop peers that are no longer in some window of peer sets like other p2p

File: p2p/src/simulated/network.rs [hunk 7]
```diff
@@ -169,9 +173,24 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
         }
 
         match message {
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs:196]: We should drop messages before attempting to send if the recipient is not in any peer set (even if there is an explicit link)

File: p2p/src/simulated/ingress.rs [hunk 8]
```diff
@@ -93,27 +104,6 @@ impl<P: PublicKey> Oracle<P> {
         r.await.map_err(|_| Error::NetworkClosed)?
     }
 
-    /// Register a new peer with the network that can interact over a given channel.
-    ///
-    /// By default, the peer will not be linked to any other peers. If a peer is already
-    /// registered on a given channel, it will return an error.
-    pub async fn register(
```
Comment (patrick-ogrady) [inline p2p/src/simulated/ingress.rs:100]: Moving this to `Control` makes a lot more sense 👍

File: p2p/src/simulated/network.rs [hunk 9]
```diff
@@ -75,6 +75,9 @@ pub struct Network<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey>
     // A map from a public key to a peer
     peers: BTreeMap<P, Peer<P>>,
 
+    // The current peer set ID
+    peer_set_id: u64,
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs]: This will likely be a BTreeSet after the requested changes (where we take a "latest")

File: resolver/src/p2p/engine.rs [hunk 10]
```diff
@@ -170,10 +180,18 @@ impl<
             self.metrics.serve_processing.set(self.serves.len() as i64);
 
             // Update peer list if-and-only-if it might have changed.
-            let peer_set_id = self.coordinator.peer_set_id();
-            if self.last_peer_set_id != Some(peer_set_id) {
-                self.last_peer_set_id = Some(peer_set_id);
-                self.fetcher.reconcile(self.coordinator.peers());
+            let peer_set_id = self.peer_provider.latest_peer_set().await;
```
Comment (patrick-ogrady) [inline resolver/src/p2p/engine.rs]: It still feels wrong to hammer this so hard (blocking channel request per loop).

Comment (patrick-ogrady) [reply resolver/src/p2p/engine.rs]: I think the practical solution here is to change this to a channel we can listen on that sends the superset of all peers we are permitted to message (based on all allowed sets).

This would be most similar to: https://github.com/commonwarexyz/monorepo/blob/2c85371f3230bb260e66798bb3caac40e08df396/consensus/src/lib.rs#L199-L212

File: p2p/src/simulated/network.rs [hunk 11]
```diff
@@ -331,6 +450,17 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
                 continue;
             }
 
+            // Check if recipient is in any tracked peer set (only if peer sets have been registered)
+            if !self.peer_sets.is_empty() && !self.peer_refs.contains_key(&recipient) {
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs]: I was wondering how you'd handle this edge case (if we should update all tests to include a notion of peer sets).

File: p2p/src/simulated/network.rs [hunk 12]
```diff
@@ -75,12 +82,24 @@ pub struct Network<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey>
...
+    tracked_peer_sets: usize,
+
     // A map of peers blocking each other
     blocks: HashSet<(P, P)>,
 
     // State of the transmitter
     transmitter: transmitter::State<P>,
 
+    // Subscribers to peer set updates
+    subscribers: Vec<mpsc::UnboundedSender<u64>>,
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs]: I think we should push `(u64, union(peer_sets))`? It seems like anyone listening to this would immediately want to get that.

File: p2p/src/simulated/network.rs [hunk 13]
```diff
@@ -169,9 +192,74 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
...
+                    *self.peer_refs.entry(public_key.clone()).or_insert(0) += 1;
+                }
+                self.peer_sets.insert(peer_set, peers);
+
+                // Notify all subscribers about the new peer set
+                self.subscribers
+                    .retain(|subscriber| subscriber.unbounded_send(peer_set).is_ok());
+
+                // Remove oldest peer set if we exceed the limit
+                while self.peer_sets.len() > self.tracked_peer_sets {
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs]: We should remove peers before notifying subscribers (when we make the change to push the union set).

File: p2p/src/simulated/network.rs [hunk 14]
```diff
@@ -42,6 +42,13 @@ pub struct Config {
     /// typically disconnect, for testing purposes it may be useful to keep peers connected,
     /// allowing byzantine actors the ability to continue sending messages.
     pub disconnect_on_block: bool,
+
+    /// The maximum number of peer sets to track. When a new peer set is registered and this
+    /// limit is exceeded, the oldest peer set is removed. Peers that are no longer in any
+    /// tracked peer set will have their links removed and messages to them will be dropped.
+    ///
+    /// Defaults to 3 if not specified, matching production p2p implementations.
+    pub tracked_peer_sets: usize,
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs]: Let's make this an `Option<usize>` for tests that don't care about peer sets (rather than implicitly assuming preferred behavior).

File: consensus/src/marshal/resolver/p2p.rs [hunk 15]
```diff
@@ -5,21 +5,21 @@ use crate::{
...
 /// Configuration for the P2P [Resolver](commonware_resolver::Resolver).
-pub struct Config<P: PublicKey, C: Coordinator<PublicKey = P>> {
+pub struct Config<P: PublicKey, C: PeerSetManager<PublicKey = P>> {
     /// The public key to identify this node.
     pub public_key: P,
 
-    /// The coordinator of peers that can be consulted for fetching data.
-    pub coordinator: C,
+    /// The provider of peers that can be consulted for fetching data.
+    pub peer_provider: C,
```
Comment (patrick-ogrady) [inline consensus/src/marshal/resolver/p2p.rs]: nit: `manager` (types already imply it has to do with `peers`)

File: examples/reshare/src/application/scheme.rs [hunk 16]
```diff
@@ -121,27 +119,3 @@ impl EpochSchemeProvider for SchemeProvider<EdScheme, ed25519::PrivateKey> {
         EdScheme::new(transition.participants.clone(), self.signer.clone())
     }
 }
-
-#[derive(Clone)]
```
Comment (patrick-ogrady) [inline examples/reshare/src/application/scheme.rs:125]: 🔥

File: examples/reshare/src/dkg/actor.rs [hunk 17]
```diff
@@ -216,6 +221,25 @@ where
             .report(orchestrator::Message::Enter(transition))
             .await;
 
+        // Register the initial set of peers.
+        let players_after_next = Ordered::from_iter(
```
Comment (patrick-ogrady) [inline examples/reshare/src/dkg/actor.rs]: can do:

```
                            Self::choose_from_all(
                                &all_participants,
                                self.num_participants_per_epoch,
                                current_epoch + 1,
                            )
                            .into_iter()
                            .collect::<Ordered<_>>()
                   ```

File: examples/reshare/src/dkg/actor.rs [hunk 18]
```diff
@@ -395,6 +419,28 @@ where
                             .collect::<Ordered<_>>()
                         };
 
+                        // Select the players for the epoch after next to allow them to sync two epochs prior
+                        // to their participation.
+                        let next_epoch_players = Ordered::from_iter(
```
Comment (patrick-ogrady) [inline examples/reshare/src/dkg/actor.rs]: see: https://github.com/commonwarexyz/monorepo/pull/1984/files#r2479021829

File: p2p/src/simulated/network.rs [hunk 19]
```diff
@@ -169,9 +193,80 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
...
+                    for public_key in set.iter() {
+                        let refs = self.peer_refs.get_mut(public_key).unwrap();
+                        *refs = refs.saturating_sub(1);
+
+                        // If peer is no longer in any tracked set, remove it
+                        if *refs == 0 {
+                            self.peer_refs.remove(public_key);
+                            self.peers.remove(public_key);
+
+                            // Remove all links involving this peer
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs]: Wonder if we should leave this as-is 🤔

File: p2p/src/simulated/network.rs [hunk 20]
```diff
@@ -202,6 +297,25 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
...
+                    let _ = response.send(self.peer_sets.get(&index).cloned());
+                }
+            }
+            ingress::Message::LatestPeerSet { response } => {
+                // Return the latest peer set or 0 if none exist
+                let _ = response.send(Some(self.peer_sets.keys().last().copied().unwrap_or(0)));
+            }
+            ingress::Message::Subscribe { response } => {
+                // Create a new subscription channel
+                let (sender, receiver) = mpsc::unbounded();
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs:309]: We should send the latest peer set here.

File: p2p/src/simulated/network.rs [hunk 21]
```diff
@@ -202,6 +297,25 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
                 );
                 send_result(result, Ok((sender, receiver)))
             }
+            ingress::Message::PeerSet { index, response } => {
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs:298]: with the subscription, do we need to support `PeerSet` and `LatestPeerSet`?

File: p2p/src/simulated/network.rs [hunk 22]
```diff
@@ -316,7 +430,19 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
         // Collect recipients
         let (channel, origin, recipients, message, reply) = task;
         let recipients = match recipients {
-            Recipients::All => self.peers.keys().cloned().collect(),
+            Recipients::All => {
+                // If peer sets have been registered, send only to tracked peers
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs:447]: We always sent to all peers (not just ones in latest)

File: p2p/src/simulated/network.rs [hunk 23]
```diff
@@ -331,6 +457,17 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
                 continue;
             }
 
+            // Check if recipient is in any tracked peer set (only if peer sets have been registered)
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs]: check tracked_sets == none

File: resolver/src/p2p/engine.rs [hunk 24]
```diff
@@ -153,8 +153,20 @@ impl<
         let (mut sender, mut receiver) = wrap((), network.0, network.1);
 
         // Set initial peer set.
-        self.last_peer_set_id = Some(self.coordinator.peer_set_id());
-        self.fetcher.reconcile(self.coordinator.peers());
+        self.last_peer_set_id = self.manager.latest_peer_set().await;
```
Comment (patrick-ogrady) [inline resolver/src/p2p/engine.rs]: as noted, can remove this by pushing latest when subscribe (what we do in the consensus version of this)

File: resolver/src/p2p/engine.rs [hunk 25]
```diff
@@ -196,6 +201,19 @@ impl<
...
                 },
 
+                // Handle peer set updates
+                peer_set_update = peer_set_subscription.next() => {
+                    let Some((id, set)) = peer_set_update else {
+                        debug!("peer set subscription closed");
+                        return;
+                    };
+
+                    if self.last_peer_set_id != Some(id) {
```
Comment (patrick-ogrady) [inline resolver/src/p2p/engine.rs]: nit: if id > last

File: p2p/src/simulated/network.rs (hunk 23 above)
Comment (clabby) [reply p2p/src/simulated/network.rs]: We don't need to; There's no way to register a peer set if `tracked_peer_sets` is `None`.

File: p2p/src/simulated/network.rs (hunk 19 above)
Comment (clabby) [reply p2p/src/simulated/network.rs]: I think we should remove all links to simulate a proper disconnection (fully). Even though the send logic still skips even if there is an active link, there's no reason to retain the link imo.

File: p2p/src/simulated/network.rs (hunk 21 above)
Comment (clabby) [reply p2p/src/simulated/network.rs:298]: Nope, removed from the trait.

File: p2p/src/authenticated/discovery/actors/tracker/actor.rs [hunk 26]
```diff
@@ -136,13 +140,29 @@ impl<E: Spawner + Rng + Clock + GClock + RuntimeMetrics, C: Signer> Actor<E, C>
                 assert!(len as u64 <= max, "peer set too large: {len} > {max}");
 
                 self.directory.add_set(index, peers);
+
+                // Notify all subscribers about the new peer set
+                let new_set = self.directory.get_set(&index).cloned().unwrap();
```
Comment (patrick-ogrady) [inline p2p/src/authenticated/discovery/actors/tracker/actor.rs]: just clone `peers`

File: p2p/src/authenticated/discovery/actors/tracker/actor.rs [hunk 27]
```diff
@@ -136,13 +140,29 @@ impl<E: Spawner + Rng + Clock + GClock + RuntimeMetrics, C: Signer> Actor<E, C>
...
                 // Send the peer set at the given index.
                 let _ = responder.send(self.directory.get_set(&index).cloned());
             }
-            Message::LatestPeerSet { responder } => {
-                let _ = responder.send(self.directory.latest_set_index());
+            Message::Subscribe { responder } => {
+                // Create a new subscription channel
+                let (sender, receiver) = mpsc::unbounded();
+
+                // Send the current peer sets immediately
```
Comment (patrick-ogrady) [inline p2p/src/authenticated/discovery/actors/tracker/actor.rs]: nit: current peer set

File: consensus/src/marshal/resolver/p2p.rs [hunk 28]
```diff
@@ -5,21 +5,21 @@ use crate::{
     Block,
 };
 use commonware_cryptography::PublicKey;
-use commonware_p2p::{utils::requester, Receiver, Sender};
-use commonware_resolver::p2p::{self, Coordinator};
+use commonware_p2p::{utils::requester, PeerSetManager, Receiver, Sender};
```
Comment (patrick-ogrady) [inline consensus/src/marshal/resolver/p2p.rs]: just `Manager`?

File: examples/reshare/src/dkg/actor.rs [hunk 29]
```diff
@@ -39,7 +39,8 @@ use tracing::info;
 
 const EPOCH_METADATA_KEY: FixedBytes<1> = fixed_bytes!("0xFF");
 
-pub struct Config<C> {
+pub struct Config<C, P> {
+    pub peer_set_manager: P,
```
Comment (patrick-ogrady) [inline examples/reshare/src/dkg/actor.rs]: nit: manager

File: examples/reshare/src/dkg/actor.rs [hunk 30]
```diff
@@ -50,14 +51,16 @@ pub struct Config<C> {
...
+pub struct Actor<E, P, H, C, V>
 where
     E: Spawner + Metrics + CryptoRngCore + Clock + GClock + Storage,
+    P: PeerSetManager<PublicKey = C::PublicKey, Peers = Ordered<C::PublicKey>>,
     H: Hasher,
     C: Signer,
     V: Variant,
 {
     context: ContextCell<E>,
+    peer_set_manager: P,
```
Comment (patrick-ogrady) [inline examples/reshare/src/dkg/actor.rs]: nit: manager

File: examples/reshare/src/dkg/actor.rs [hunk 31]
```diff
@@ -216,6 +221,22 @@ where
             .report(orchestrator::Message::Enter(transition))
             .await;
 
+        // Register the initial set of peers.
+        let players_after_next = Self::choose_from_all(
+            &all_participants,
+            self.num_participants_per_epoch,
+            current_epoch + 1,
+        );
+        let next_epoch_peers = dealers
```
Comment (patrick-ogrady) [inline examples/reshare/src/dkg/actor.rs]: nit: we don't need to do this union because we keep some number of sets

File: examples/reshare/src/dkg/actor.rs [hunk 32]
```diff
@@ -395,6 +416,25 @@ where
...
+                        // Select the players for the epoch after next to allow them to sync two epochs prior
+                        // to their participation.
+                        let next_epoch_players = Self::choose_from_all(
+                            &all_participants,
+                            self.num_participants_per_epoch,
+                            next_epoch + 1,
+                        );
+
+                        // Register the set of peers for the next epoch.
+                        let next_epoch_peers = next_participants
```
Comment (patrick-ogrady) [inline examples/reshare/src/dkg/actor.rs]: nit: we don't need to do this union

File: examples/reshare/src/engine.rs [hunk 33]
```diff
@@ -40,15 +40,17 @@ const REPLAY_BUFFER: NonZero<usize> = NZUsize!(8 * 1024 * 1024); // 8MB
...
-pub struct Config<C, B, V>
+pub struct Config<C, P, B, V>
 where
     C: Signer,
+    P: PeerSetManager<PublicKey = C::PublicKey, Peers = Ordered<C::PublicKey>>,
     B: Blocker<PublicKey = C::PublicKey>,
     V: Variant,
 {
     pub signer: C,
+    pub peer_set_manager: P,
```
Comment (patrick-ogrady) [inline examples/reshare/src/engine.rs]: nit: manager

File: examples/reshare/src/validator.rs [hunk 34]
```diff
@@ -118,11 +114,12 @@ where
     };
     let marshal = marshal_resolver::init(&context, resolver_cfg, marshal);
 
-    let engine = engine::Engine::<_, _, _, Sha256, MinSig, S>::new(
+    let engine = engine::Engine::<_, _, _, _, Sha256, MinSig, S>::new(
         context.with_label("engine"),
         engine::Config {
             signer: config.signing_key.clone(),
-            blocker: oracle,
+            peer_set_manager: oracle.clone(),
```
Comment (patrick-ogrady) [inline examples/reshare/src/validator.rs]: nit: manager

File: examples/reshare/src/validator.rs [hunk 35]
```diff
@@ -205,42 +202,23 @@ mod test {
         ),
     > {
         let mut registrations = HashMap::new();
-        let ordered_validators = validators.iter().cloned().collect::<Ordered<_>>();
         for validator in validators.iter() {
-            let pending = oracle
-                .register(validator.clone(), PENDING_CHANNEL)
-                .await
-                .unwrap();
+            let mut control = oracle.control(validator.clone());
```
Comment (patrick-ogrady) [inline examples/reshare/src/validator.rs:206]: ❤️

File: p2p/src/simulated/ingress.rs [hunk 36]
```diff
@@ -10,11 +11,22 @@ use std::time::Duration;
 
 pub enum Message<P: PublicKey> {
     Register {
-        public_key: P,
+        peer_set: u64,
+        peers: Ordered<P>,
+    },
+    RegisterComms {
```
Comment (patrick-ogrady) [inline p2p/src/simulated/ingress.rs]: nit: `Register`

File: p2p/src/simulated/ingress.rs [hunk 37]
```diff
@@ -10,11 +11,22 @@ use std::time::Duration;
 
 pub enum Message<P: PublicKey> {
     Register {
```
Comment (patrick-ogrady) [inline p2p/src/simulated/ingress.rs]: nit: `Update`

File: resolver/src/p2p/engine.rs [hunk 38]
```diff
@@ -196,6 +186,19 @@ impl<
                     return;
                 },
 
+                // Handle peer set updates
+                peer_set_update = peer_set_subscription.next() => {
```
Comment (patrick-ogrady) [inline resolver/src/p2p/engine.rs:190]: If we just have the most recent set, we may only fetch from new/syncing peers

File: p2p/src/simulated/network.rs (hunk 23 above)
Comment (patrick-ogrady) [reply p2p/src/simulated/network.rs]: I'm concerned about the race on initial registration (I consider it different if no peer sets have been registered and no peer sets have been registered yet)

File: p2p/src/simulated/network.rs [hunk 39]
```diff
@@ -316,7 +429,19 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
...
         let recipients = match recipients {
-            Recipients::All => self.peers.keys().cloned().collect(),
+            Recipients::All => {
+                // If peer sets have been registered, send only to tracked peers
+                // Otherwise, send to all registered peers (compatibility
+                // with tests that do not register peer sets.)
+                if self.peer_sets.is_empty() {
+                    self.peers.keys().cloned().collect()
+                } else {
+                    // Sort keys for deterministic ordering (peer_refs is a HashMap)
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs]: nit: just store as a BTreeMap (we'll call this a TON)

File: p2p/src/authenticated/discovery/actors/tracker/actor.rs [hunk 40]
```diff
@@ -134,8 +139,33 @@ impl<E: Spawner + Rng + Clock + GClock + RuntimeMetrics, C: Signer> Actor<E, C>
...
+            }
+            Message::Subscribe { responder } => {
+                // Create a new subscription channel
+                let (sender, receiver) = mpsc::unbounded();
+
+                // Send the latest peer set immediately
+                if let Some(latest_set_id) = self.directory.latest_set_index() {
+                    let latest_set = self.directory.get_set(&latest_set_id).cloned().unwrap();
+                    sender
+                        .unbounded_send((latest_set_id, latest_set, self.directory.tracked()))
```
Comment (patrick-ogrady) [inline p2p/src/authenticated/discovery/actors/tracker/actor.rs:163]: While this may seem like a lot of data to send here, keep in mind there should only be a handful of listeners and this will change every ~week

File: p2p/src/authenticated/lookup/actors/tracker/directory.rs [hunk 41]
```diff
@@ -135,7 +135,7 @@ impl<E: Spawner + Rng + Clock + GClock + RuntimeMetrics, C: PublicKey> Directory
             };
             record.increment();
         }
-        let peers: Vec<_> = peers.into_iter().map(|(peer, _)| peer).collect();
+        let peers: Ordered<_> = peers.into_iter().map(|(peer, _)| peer).collect();
```
Comment (patrick-ogrady) [inline p2p/src/authenticated/lookup/actors/tracker/directory.rs]: nit: peers.keys()

File: p2p/src/simulated/network.rs [hunk 42]
```diff
@@ -169,9 +194,77 @@ impl<E: RNetwork + Spawner + Rng + Clock + Metrics, P: PublicKey> Network<E, P>
...
+
+                // Remove oldest peer set if we exceed the limit
+                while self.peer_sets.len() > tracked_peer_sets {
+                    let (index, set) = self.peer_sets.pop_first().unwrap();
+                    debug!(index, "removed oldest peer set");
+
+                    // Decrement reference counts and clean up peers/links
+                    for public_key in set.iter() {
+                        let refs = self.peer_refs.get_mut(public_key).unwrap();
+                        *refs = refs.saturating_sub(1);
```
Comment (patrick-ogrady) [inline p2p/src/simulated/network.rs]: This "should" never go below 0 if correct, we should probably make this a checked_sub.
